modules/lang-mustache/src/main/java/org/elasticsearch/rest/action/search/template/RestMultiSearchTemplateAction.java

```diff
@@ -0,0 +1,205 @@
+import org.elasticsearch.search.suggest.Suggesters;
+
+import java.util.Map;
+
+import static org.elasticsearch.common.xcontent.support.XContentMapValues.lenientNodeBooleanValue;
+import static org.elasticsearch.common.xcontent.support.XContentMapValues.nodeStringArrayValue;
+import static org.elasticsearch.common.xcontent.support.XContentMapValues.nodeStringValue;
+import static org.elasticsearch.rest.RestRequest.Method.GET;
+import static org.elasticsearch.rest.RestRequest.Method.POST;
+
+public class RestMultiSearchTemplateAction extends BaseRestHandler {
+
+    private final boolean allowExplicitIndex;
+    private final IndicesQueriesRegistry queryRegistry;
+    private final AggregatorParsers aggParsers;
+    private final Suggesters suggesters;
+
+    @Inject
+    public RestMultiSearchTemplateAction(Settings settings, RestController controller, Client client, IndicesQueriesRegistry queryRegistry,
+                                         AggregatorParsers aggregatorParsers, Suggesters suggesters) {
+        super(settings, client);
+        this.queryRegistry = queryRegistry;
+        this.aggParsers = aggregatorParsers;
+        this.suggesters = suggesters;
+        this.allowExplicitIndex = MULTI_ALLOW_EXPLICIT_INDEX.get(settings);
+
+        controller.registerHandler(GET, "/_msearch/template", this);
+        controller.registerHandler(POST, "/_msearch/template", this);
+        controller.registerHandler(GET, "/{index}/_msearch/template", this);
+        controller.registerHandler(POST, "/{index}/_msearch/template", this);
+        controller.registerHandler(GET, "/{index}/{type}/_msearch/template", this);
+        controller.registerHandler(POST, "/{index}/{type}/_msearch/template", this);
+    }
+
+    @Override
+    protected void handleRequest(RestRequest request, RestChannel channel, Client client) throws Exception {
+        if (RestActions.hasBodyContent(request) == false) {
+            throw new ElasticsearchException("request body is required");
+        }
+
+        MultiSearchTemplateRequest multiRequest = new MultiSearchTemplateRequest();
+
+        String[] indices = Strings.splitStringByCommaToArray(request.param("index"));
+        String[] types = Strings.splitStringByCommaToArray(request.param("type"));
+        IndicesOptions indicesOptions = IndicesOptions.fromRequest(request, multiRequest.indicesOptions());
+        String routing = request.param("routing");
+        String searchType = request.param("search_type");
+
+        parseRequest(multiRequest, RestActions.getRestContent(request), indices, types, searchType, routing, indicesOptions,
+                allowExplicitIndex, queryRegistry, parseFieldMatcher, aggParsers, suggesters);
+
+        client.execute(MultiSearchTemplateAction.INSTANCE, multiRequest, new RestToXContentListener<>(channel));
+    }
+
+    public static MultiSearchTemplateRequest parseRequest(MultiSearchTemplateRequest multiSearchRequest, BytesReference data,
```

Comment:
I think we should try to come up with a general static method for both `RestMultiSearchTemplateAction` and `RestMultiSearchAction`? The shared helper method could have Function<BytesReference, Request> parameter, so that each action can reuse the logic while returning its own classes.